README.md
<p align="center">
<a href="https://github.com/GentleCP/corpwechat-bot"><img width="300" src="https://gitee.com/gentlecp/ImgUrl/raw/master/20210425111523.png"></a>
</p>

<p align="center">
<a href="https://hits.seeyoufarm.com"><img src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2FGentleCP%2Fcorpwechat-bot&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false"/></a>
<a ><img src="https://img.shields.io/badge/python-3.5%2B-blue"/></a>
<a ><img src="https://img.shields.io/pypi/v/corpwechatbot"/></a>
</p>

👋 `corpwechatbot`是一个`python`封装的企业机器人&应用消息推送库，通过企业微信提供的api实现。

利用本库，你可以轻松地实现从服务器端发送一条文本、图片、视频、`markdown`等等消息到你的微信手机端，而不依赖于其他的第三方应用

![](docs/img/work_mechanism.png)

![](docs/img/app_msgsend.gif)
## Usage
[点此查看说明文档](https://corpwechatbot.gentlecp.com) ，若无法打开，也可直接访问[原始文档](docs/index.md)

## Author

[@GentleCP](https://github.com/GentleCP)

## License

本项目遵守[GPL v3](LICENSE)开源协议

##  Sponsor

本项目完全出于公益性质开源使用，作者独自完成开发测试等一系列工作，如果它帮到了你，解决了你的痛点需求，你可以请作者喝杯阔乐，或者给这个项目一个`star`（右上角），激励作者继续更新完善本项目，非常感谢🙏

![](https://cdn.jsdelivr.net/gh/GentleCP/ImgUrl/1411624691159_.pic_hd.jpg)



docs/changelog.md
# 更新日志
## v0.6.4: 22/08/29
- `fix`: 修复发送markdown消息时文件名过长导致的问题，现在仅读取`.md`结尾的文件，若文件不存在，则直接发送内容
## v0.6.3: 22/08/02
- `feat`: 现在支持在创建`AppMsgSender,CorpWechatBot`时提供`proxies`参数，通过代理发送消息，适合内网环境中无法直接连接外网的情况
## v0.6.2: 22/03/05
- `feat`: 现在创建群聊`app.create_chat(show_chat=True)`的时候可设置`show_chat`选项，表是否在群聊成功创建后发送第一条消息到群聊，让群聊显示在会话列表中，发送的消息会包含群聊id信息。

## v0.6.1: 21/08/22
- `fix`: 解决[issue](https://github.com/GentleCP/corpwechatbot/issues/9)
- `feat`: 支持用户自行决定是否显示日志，在初始化`AppMsgSender`,`CorpwechatBot`实例的时候传入`log_level`参数决定日志的显示等级，可选等级有：`50(CRITICAL), 40(ERROR), 30(WARNING), 20(INFO,默认), 10(DEBUG), 0(NOTSET)`, 下面是一个例子：
```python
import cptools
from corpwechatbot import AppMsgSender

app = AppMsgSender(log_level=20)  # 显示INFO及以上级别的log
app2 = AppMsgSender(log_level=cptools.INFO)  # 你也可以选择易读性更高的传入方式
```

## v0.6.0: 21/08/05
- `feat`: 支持应用群聊创建与消息推送
- `feat`: 支持多个本地文件key，用于存储多个密钥信息
- `refactor`: 优化重构部分代码，移除多余文件
- `doc`: 优化文档内容

## v0.5.2: 21/07/19
- `fix`: 修复token过期导致发送消息失败的问题，当程序长时间运行导致获取的token过期时，会自动更新token，并重新发送一遍失败的消息
![](https://cdn.jsdelivr.net/gh/GentleCP/ImgUrl/20210719204916.png)
## v0.5.1: 21/07/18
- `refactor`: 重构部分代码，提升简洁性
- `fix`: ~~修复token过期导致发送失败的问题（程序运行超过2小时）~~
> 经检查，本次修复的实际上是无效token的问题，并没解决token过期的问题

## v0.5.0: 21/06/27
- `feat`: 添加对任务卡片消息发送的支持
- `feat`: 添加服务器回调配置程序
- `docs`: 简化说明文档

## v0.4.0: 21/05/21
- `feat`: 添加对`mpnews`类型消息的支持，关于该类型的解释参考[官网](https://work.weixin.qq.com/api/doc/90000/90135/90236#%E5%9B%BE%E6%96%87%E6%B6%88%E6%81%AF%EF%BC%88mpnews%EF%BC%89)
- `feat`: 添加对参数`enable_id_trans`, `enable_duplicate_check`,`duplicate_check_interval`，现在你可以在发送消息的时候指定它们，具体参考官网
- `refactor`: 更改参数`safe`为`int`，含义与官网等同，优化部分代码

## v0.3.1: 21/05/17
- `fix`: 修复不同企业同一id应用的token存储冲突问题
## v0.3.0: 21/05/16
-  `feat`: 支持在同一个项目中创建多个应用消息推送

## v0.2.4: 21/05/11
- `fix`：修复因单例模式参数设置不正确导致实例化`AppMsgSender`异常的问题
- `v0.2.1`特性继承
- `v0.2.2`特性继承
- `v0.2.3`特性继承

## ~~v0.2.3: 21/05/10~~
- `fix`: 修复从`pypi`安装`corpwechatbot`出现的找不到`pypidoc.md`的错误

## ~~v0.2.2: 21/05/07~~

- `fix`: 1000004
- `refactor`: 优化重构`AppMsgSender`部分代码，更改传参机制
- `docs`: 更新使用文档和README文档的内容

## ~~v0.2.1: 21/04/24~~
- `fix`: 修复终端直接运行-u参数随意输入报错的问题
- `feat`: 现在支持在发送`markdown`信息的时候直接传入`markdown`文件路径
- `docs`: 添加专用用于`pypi`的文档
- `docs`: 添加为何使用`corpwechatbot`的文档说明到`README`

## v0.2.0: 21/04/18
- `feat`: 允许在本地创建文件读取企业微信配置，避免每次程序都要传入
- `feat`: 支持在命令行直接执行包发送信息，如`cwb -u='app' -t='hello world'`
- `fix`：token存储位置修改，初始设置保存到本地`site-packages`（后期将其移到`site-packages/corpwechatbot`目录下，方便统一）

## v0.1.0: 21/04/09
- 项目初始化建立
- 添加群聊机器人消息推送
- 添加应用消息推送，暂不支持**小程序消息、任务卡片消息**

corpwechatbot/app.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
-----------------File Info-----------------------
            Name: app.py
            Description: 企业微信应用消息推送
            Author: GentleCP
            Email: me@gentlecp.com
            WebSite: https://blog.gentlecp.com
            Create Date: 2021/4/7
-----------------End-----------------------------
"""
import requests
import json
import hashlib
from typing import Optional
from datetime import datetime
from pathlib import Path
from typing import List, Dict

from corpwechatbot._sender import MsgSender
from corpwechatbot.error import TokenGetError
from corpwechatbot.util import is_image, is_voice, is_video, is_file
from corpwechatbot.config import OFFICIAL_APIS

CUR_PATH = Path(__file__)
TOKEN_PATH = CUR_PATH.parent.joinpath('token.json')  # 存储在本项目根目录下


class AppMsgSender(MsgSender):
    """
    应用消息推送器，支持文本、图片、语音、视频、文件、文本卡片、图文、markdown消息推送
    """

    def __init__(self,
                 corpid: str = '',
                 corpsecret: str = '',
                 agentid: str = '',
                 log_level: int = 20,
                 **kwargs):
        """
        :param corpid: 企业id
        :param corpsecret: 应用密钥
        :param agentid: 应用id
        """
        super().__init__(log_level, **kwargs)
        corpkeys = self._get_corpkeys(corpid=corpid, corpsecret=corpsecret, agentid=agentid, **kwargs)
        self.corpid = corpkeys.get('corpid', '')
        self.corpsecret = corpkeys.get('corpsecret', '')
        self.agentid = corpkeys.get('agentid', '')
        self._token_key = hashlib.sha1(bytes(self.corpid + self.agentid, encoding='utf-8')).hexdigest()
        # APIs
        self.get_token_api = self.base_url.format(
            OFFICIAL_APIS['GET_ACCESS_TOKEN'].format(self.corpid, self.corpsecret))
        self.__fresh_msg_send_api()
        self.appchat_create_api = self.base_url.format(OFFICIAL_APIS['APPCHAT_CREATE'].format(self.access_token))
        self.appchat_send_api = self.base_url.format(OFFICIAL_APIS['APPCHAT_SEND'].format(self.access_token))

    def __fresh_msg_send_api(self):
        self.access_token = self.get_assess_token()
        self.msg_send_api = self.base_url.format(OFFICIAL_APIS['MESSAGE_SEND'].format(self.access_token))

    def _get_corpkeys(self, corpid: str = '', corpsecret: str = '', agentid: str = '', **kwargs):
        """
        get keys for app from parameter or local
        :param corpid:
        :param corpsecret:
        :param agentid:
        :return:
        """
        if corpid and corpsecret and agentid:
            return {
                'corpid': corpid,
                'corpsecret': corpsecret,
                'agentid': agentid,
            }
        else:
            # from local
            res = {}
            options = ['corpid', 'corpsecret', 'agentid']
            for k, v in zip(options, self._get_local_keys(section='app', options=options, **kwargs)):
                res.update({k: v})
            return res

    def get_assess_token(self):
        """
        通过企业id和应用凭证密钥获取assess_token，用于消息推送，根据corpid和agentid识别不同的token
        """
        try:
            token_dict = json.loads(TOKEN_PATH.read_text())
        except FileNotFoundError:
            # 尚未获取过token
            self.logger.debug('旧token获取失败，重新获取token')
            return self.__get_access_token()
        else:
            try:
                token_info = token_dict[self._token_key]
            except KeyError:
                # 该agentid对应token不存在
                self.logger.debug("token不存在，获取token")
                return self.__get_access_token(token_dict)
            else:
                if token_info.get('expire_time', float("-inf")) < datetime.now().timestamp():
                    self.logger.debug("token过期，重新获取token")
                    return self.__get_access_token(token_dict)
                return token_info['token']

    def __get_access_token(self, token_dict={}):
        res = requests.get(self.get_token_api, proxies=self.proxies).json()
        if res.get('errcode') == 0:
            self.logger.info("token请求成功")
            token = res.get('access_token')
            token_dict.update({
                self._token_key: {
                    'token': token,
                    'expire_time': datetime.now().timestamp() + 7200
                }
            })
            TOKEN_PATH.write_text(json.dumps(token_dict))
            return token
        else:
            raise TokenGetError(f"token请求失败，原因：{res.get('errmsg', '无法获取token')}")

    def __list2str(self, datas: []):
        """
        将传入的list数据转换成 | 划分的字符串
        e.g. ['user1', 'user2'] -> 'user1|user2'
        :param datas:
        :return:
        """
        return "".join([item + '|' for item in datas])[:-1]

    def __check_type_and_send(self, msg_type, data, media_path):
        """
        :param msg_type:
        :param data:
        :param media_path:
        :return:
        """
        media_types = {'image', 'voice', 'video', 'file'}
        if msg_type in media_types:
            # 发送媒体消息
            self._media_api = self.base_url.format(OFFICIAL_APIS['MEDIA_UPLOAD'].format(self.access_token, msg_type))
            media_res = self._get_media_id(media_type=msg_type, p_media=Path(media_path))
            data[msg_type] = {
                "media_id": media_res.get('media_id', '')
            }
        elif msg_type == 'mpnews':
            # mpnews比较特殊，单独处理
            self._media_api = self.base_url.format(OFFICIAL_APIS['MEDIA_UPLOAD'].format(self.access_token, 'image'))
            media_res = self._get_media_id(media_type='image', p_media=Path(media_path))
            data[msg_type]["articles"][0]["thumb_media_id"] = media_res.get('media_id', None)
        if data['chatid']:
            return self._post(self.appchat_send_api, data)
        return self._post(self.msg_send_api, data)

    def _send(self,
              msg_type: str = '',
              data: dict = {},
              media_path: Optional[str] = '',
              **kwargs):
        """
        新的统一内部发送接口，供不同消息推送接口调用
        :param msg_type:
        :param data:
        :param media_path: 只有需要media id的时候才传入
        :param kwargs:
        :return:
        """
        # prepare data
        if not (kwargs.get('touser') or kwargs.get('toparty') or kwargs.get('totag')):
            # 三者均为空，默认发送全体成员
            kwargs['touser'] = ['@all']
        data['chatid'] = kwargs.get('chatid', '')
        if not data['chatid']:
            # 不是发送到群聊的消息
            data.update({
                "touser": self.__list2str(kwargs.get('touser', [])),
                "toparty": self.__list2str(kwargs.get('toparty', [])),
                "totag": self.__list2str(kwargs.get('totag', [])),
                "agentid": self.agentid,
                "enable_id_trans": kwargs.get('enable_id_trans'),
                "enable_duplicate_check": kwargs.get('enable_duplicate_check'),
                "duplicate_check_interval": kwargs.get('duplicate_check_interval')
            })
        data.update({
            "msgtype": msg_type,  # 注意这里msg_type已经是正确了，后面的改动不影响
            "safe": kwargs.get('safe', 0)
        })
        # 检查消息类型是否需要获取media_id, 如需要，则获取，并做相应检查
        send_res = self.__check_type_and_send(msg_type, data, media_path)
        if send_res.get('errcode') == 0:
            return send_res
        elif send_res.get('errcode') == 40014 or send_res.get('errcode') == 42001:
            # invalid access token or token expired, refresh token
            self.logger.info("尝试重新获取token并发送消息")
            self.__fresh_msg_send_api()
            return self.__check_type_and_send(msg_type, data, media_path)
        else:
            self.logger.error(f"发送失败! 原因：{send_res['errmsg']}")
            return send_res

    def send_image(self,
                   image_path: str,
                   **kwargs):
        """
        发送图片，支持jpg、png、bmp
        :param image_path: 图片存储路径
        :return:
        """
        if not is_image(image_path):
            self.logger.error(self.errmsgs['image_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['image_error']
            }
        return self._send(msg_type='image', media_path=image_path, **kwargs)

    def send_voice(self,
                   voice_path: str,
                   **kwargs):
        """
        发送语音，2MB，播放长度不超过60s，仅支持AMR格式
        :param voice_path:
        :return:
        """
        if not is_voice(voice_path):
            self.logger.error(self.errmsgs['voice_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['voice_error']
            }
        return self._send(msg_type='voice', media_path=voice_path, **kwargs)

    def send_video(self,
                   video_path: str,
                   **kwargs):
        """
        发送视频
        :param video_path:
        :return:
        """
        if not is_video(video_path):
            self.logger.error(self.errmsgs['video_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['video_error']
            }
        return self._send(msg_type='video', media_path=video_path, **kwargs)

    def send_file(self,
                  file_path: str,
                  **kwargs):
        """
        发送文件
        :param file_path:
        :return:
        """
        if not is_file(file_path):
            self.logger.error(self.errmsgs['file_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['file_error']
            }
        return self._send(msg_type='file', media_path=file_path, **kwargs)

    def send_text(self,
                  content: str,
                  **kwargs):
        """
        发送text消息
        :param content: 消息内容，最长不超过2048个字节，超过将
        :param safe: 是否是保密消息，False表示可对外分享，True表示不能分享且内容显示水印，默认为False，下面方法同，不再重复解释
        :param kwargs: touser, toparty, totag
        :return: send result
        """
        if not content:
            self.logger.error(self.errmsgs['text_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['text_error']
            }
        data = {
            "text": {
                "content": content
            },
        }
        return self._send(msg_type='text', data=data, **kwargs)

    def send_news(self,
                  title: str,
                  desp: Optional[str],
                  url: str,
                  picurl: Optional[str],
                  **kwargs):
        """
        发送图文消息
        :param title: 图文标题，不超过128个字节，超过会自动截断
        :param desp: 图文描述，可选，不超过512个字节，超过会自动截断
        :param url: 跳转链接
        :param picurl: 图片url，支持JPG、PNG格式，较好的效果为大图 1068*455，小图150*150。
        :return:
        """
        if not (title and url):
            self.logger.error(self.errmsgs['news_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['news_error']
            }
        data = {
            "news": {
                "articles": [
                    {
                        "title": title,
                        "description": desp,
                        "url": url,
                        "picurl": picurl
                    }
                ]
            },
        }
        return self._send(msg_type='news', data=data, **kwargs)

    def send_mpnews(self,
                    title: str,
                    image_path: str,
                    content: str,
                    author: Optional[str],
                    content_source_url: Optional[str],
                    digest: Optional[str],
                    **kwargs):
        """
        发送mpnews消息
        :param title: 图文标题
        :param image_path: 缩略图所在路径
        :param content: 图文消息内容
        :param author: 作者信息
        :param content_source_url: 点击跳转链接
        :param digest: 图文消息描述
        :param kwargs:
        :return:
        """
        if not (title and image_path and content):
            self.logger.error(self.errmsgs['mpnews_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['mpnews_error']
            }
        data = {
            "mpnews": {
                "articles": [
                    {
                        "title": title,
                        "author": author,
                        "content_source_url": content_source_url,
                        "content": content,
                        "digest": digest
                    }
                ]
            }}
        return self._send(msg_type='mpnews', data=data, media_path=image_path, **kwargs)

    def send_markdown(self,
                      content: str,
                      **kwargs):
        """
        发送markdown消息
        :param content: markdown文本数据或markdown文件路径
        :return:
        """
        if not content:
            self.logger.error(self.errmsgs['markdown_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['markdown_error']
            }
        if content.endswith('.md'):
            try:
                content = Path(content).read_text()
            except (OSError, FileNotFoundError) as e:
                self.logger.warning('你可能正在尝试发送一个markdown文件，但文件并不存在或文件名过长，将直接发送内容.')
        data = {
            "markdown": {
                "content": content,
            },
        }
        return self._send(msg_type='markdown', data=data, **kwargs)

    def send_card(self,
                  title: str,
                  desp: str,
                  url: str,
                  btntxt: Optional[str],
                  **kwargs):
        """
        发送卡片消息
        :param title: 标题，不超过128个字节，超过会自动截断
        :param desp: 描述，不超过512个字节，超过会自动截断
        :param url: 点击后跳转的链接
        :param btntxt: 按钮文字。 默认为“详情”， 不超过4个文字，超过自动截断
        :return:
        """
        if not (title and desp and url):
            self.logger.error(self.errmsgs['card_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['card_error']
            }
        data = {
            "textcard": {
                "title": title,
                "description": desp,
                "url": url,
                "btntxt": btntxt
            },
        }
        return self._send(msg_type='textcard', data=data, **kwargs)

    def send_taskcard(self,
                      title: str,
                      desp: Optional[str],
                      url: str,
                      task_id: str,
                      btn: List[Dict],
                      **kwargs):
        """
        发送任务发片消息
        :param title: 标题，不超过128个字节，超过会自动截断（支持id转译）
        :param desp: 描述，不超过512个字节，超过会自动截断（支持id转译）
        :param url: 点击后跳转的链接。最长2048字节，请确保包含了协议头(http/https)，需要先自行配置回调服务
        :param task_id: 任务id，同一个应用发送的任务卡片消息的任务id不能重复，只能由数字、字母和“_-@”组成，最长支持128字节
        :param btn: 按钮列表，按钮个数为1~2个，btn的例子如下，各个具体参数的含义参考：https://open.work.weixin.qq.com/api/doc/90000/90135/90236#%E4%BB%BB%E5%8A%A1%E5%8D%A1%E7%89%87%E6%B6%88%E6%81%AF
        "btn":[
            {
                "key": "key111",
                "name": "批准",
                "color":"red",
                "is_bold": true
            },
            {
                "key": "key222",
                "name": "驳回"
            }
        ]
        :param kwargs: 其他的通用参数，在此统一处理
        -------接收方的数据----
        {
           'ToUserName': 企业号,
           'FromUserName': 发送消息的用户名
           'MsgType': 'event',
           'CreateTime': '1624762869',
           'EventKey': 事件key，服务端根据key值确定用户点击的情况
           'TaskId': 任务id，这个必须是唯一的，在同一应用中发过之后不能重复(建议通过时间戳hash结合的方式生成)
           'Agentid': 应用id
        }
        :return:
        """
        if not (title and task_id and btn):
            self.logger.error(self.errmsgs['taskcard_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['taskcard_error']
            }
        data = {
            "interactive_taskcard": {
                "title": title,
                "description": desp,
                "url": url,
                "task_id": task_id,
                "btn": btn,
            },
        }
        return self._send(msg_type='interactive_taskcard', data=data, **kwargs)

    def create_chat(self,
                    users: list,
                    name: Optional[str] = '',
                    owner: Optional[str] = '',
                    chatid: Optional[str] = '',
                    show_chat: Optional[bool] = True):
        """
        创建应用群聊
        :param users: 用户id列表，至少2人
        :param name: 群聊名称
        :param owner: 群主id，不指定会随机
        :param chatid:
        :param show_chat: 群聊创建成功后是否发送一条消息让群聊在列表中显示出来（默认发送）
        :return:
        """
        if len(users) < 2:
            self.logger.error(self.errmsgs['create_chat_error'])
            return {
                'errcode': 404,
                'errmsg': self.errmsgs['create_chat_error']
            }
        data = {
            "name": name,
            "owner": owner,
            "userlist": users,
            "chatid": chatid,
        }
        res = self._post(url=self.appchat_create_api, data=data)
        if show_chat and res['errcode'] == 0:
            self.logger.info('群聊创建成功，发送第一条消息')
            self.send_text(content=f"群聊已创建，本群聊id为：{res['chatid']}", chatid=res['chatid'])
            return res
        else:
            return res

